// rsuite/esm/SelectPicker/ListboxOptionGroup.js
import _extends from "@babel/runtime/helpers/esm/extends";
import _taggedTemplateLiteralLoose from "@babel/runtime/helpers/esm/taggedTemplateLiteralLoose";
import _objectWithoutPropertiesLoose from "@babel/runtime/helpers/esm/objectWithoutPropertiesLoose";

var _templateObject, _templateObject2;

import React from 'react';
import classNames from 'classnames';
import { useClassNames } from '../utils';
import ArrowDown from '@rsuite/icons/legacy/ArrowDown';
import useUniqueId from '../utils/useUniqueId';
var ListboxOptionGroup = /*#__PURE__*/React.forwardRef(function (props, ref) {
  var _props$classPrefix = props.classPrefix,
      classPrefix = _props$classPrefix === void 0 ? 'dropdown-menu-group' : _props$classPrefix,
      title = props.title,
      children = props.children,
      className = props.className,
      _props$expanded = props.expanded,
      expanded = _props$expanded === void 0 ? true : _props$expanded,
      onClickTitle = props.onClickTitle,
      rest = _objectWithoutPropertiesLoose(props, ["classPrefix", "title", "children", "className", "expanded", "onClickTitle"]);

  var _useClassNames = useClassNames(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      prefix = _useClassNames.prefix,
      merge = _useClassNames.merge;

  var classes = merge(className, withClassPrefix());
  var groupId = useUniqueId('listbox-group-');
  var labelId = groupId + '-label';
  return /*#__PURE__*/React.createElement("div", _extends({
    role: "group",
    id: groupId,
    "aria-expanded": expanded,
    "aria-labelledby": labelId
  }, rest, {
    ref: ref,
    className: classNames(classes, {
      folded: !expanded
    })
  }), /*#__PURE__*/React.createElement("div", {
    className: prefix(_templateObject || (_templateObject = _taggedTemplateLiteralLoose(["title"]))),
    tabIndex: -1,
    onClick: onClickTitle
  }, /*#__PURE__*/React.createElement("span", {
    id: labelId
  }, title), /*#__PURE__*/React.createElement(ArrowDown, {
    "aria-hidden": true,
    className: prefix(_templateObject2 || (_templateObject2 = _taggedTemplateLiteralLoose(["caret"])))
  })), children);
});
ListboxOptionGroup.displayName = 'Listbox.OptionGroup';
export default ListboxOptionGroup;
