// rsuite/cjs/SelectPicker/ListboxOptionGroup.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _taggedTemplateLiteralLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/taggedTemplateLiteralLoose"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _classnames = _interopRequireDefault(require("classnames"));

var _utils = require("../utils");

var _ArrowDown = _interopRequireDefault(require("@rsuite/icons/legacy/ArrowDown"));

var _useUniqueId = _interopRequireDefault(require("../utils/useUniqueId"));

var _templateObject, _templateObject2;

var ListboxOptionGroup = /*#__PURE__*/_react.default.forwardRef(function (props, ref) {
  var _props$classPrefix = props.classPrefix,
      classPrefix = _props$classPrefix === void 0 ? 'dropdown-menu-group' : _props$classPrefix,
      title = props.title,
      children = props.children,
      className = props.className,
      _props$expanded = props.expanded,
      expanded = _props$expanded === void 0 ? true : _props$expanded,
      onClickTitle = props.onClickTitle,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["classPrefix", "title", "children", "className", "expanded", "onClickTitle"]);

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix,
      prefix = _useClassNames.prefix,
      merge = _useClassNames.merge;

  var classes = merge(className, withClassPrefix());
  var groupId = (0, _useUniqueId.default)('listbox-group-');
  var labelId = groupId + '-label';
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    role: "group",
    id: groupId,
    "aria-expanded": expanded,
    "aria-labelledby": labelId
  }, rest, {
    ref: ref,
    className: (0, _classnames.default)(classes, {
      folded: !expanded
    })
  }), /*#__PURE__*/_react.default.createElement("div", {
    className: prefix(_templateObject || (_templateObject = (0, _taggedTemplateLiteralLoose2.default)(["title"]))),
    tabIndex: -1,
    onClick: onClickTitle
  }, /*#__PURE__*/_react.default.createElement("span", {
    id: labelId
  }, title), /*#__PURE__*/_react.default.createElement(_ArrowDown.default, {
    "aria-hidden": true,
    className: prefix(_templateObject2 || (_templateObject2 = (0, _taggedTemplateLiteralLoose2.default)(["caret"])))
  })), children);
});

ListboxOptionGroup.displayName = 'Listbox.OptionGroup';
var _default = ListboxOptionGroup;
exports.default = _default;
