// rsuite/cjs/SelectPicker/ListboxOption.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

exports.__esModule = true;
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutPropertiesLoose2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutPropertiesLoose"));

var _react = _interopRequireDefault(require("react"));

var _utils = require("../utils");

var ListboxOption = /*#__PURE__*/_react.default.forwardRef(function ListboxOption(props, ref) {
  var selected = props.selected,
      _props$classPrefix = props.classPrefix,
      classPrefix = _props$classPrefix === void 0 ? 'dropdown-menu-item' : _props$classPrefix,
      children = props.children,
      className = props.className,
      disabled = props.disabled,
      active = props.active,
      onKeyDown = props.onKeyDown,
      rest = (0, _objectWithoutPropertiesLoose2.default)(props, ["selected", "classPrefix", "children", "className", "disabled", "active", "onKeyDown"]);

  var _useClassNames = (0, _utils.useClassNames)(classPrefix),
      withClassPrefix = _useClassNames.withClassPrefix;

  var classes = withClassPrefix({
    active: selected,
    focus: active,
    disabled: disabled
  });
  return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({
    ref: ref,
    role: "option",
    "aria-selected": selected || undefined,
    "aria-disabled": disabled
  }, rest, {
    className: className,
    onKeyDown: disabled ? undefined : onKeyDown
  }), /*#__PURE__*/_react.default.createElement("span", {
    className: classes
  }, children));
});

ListboxOption.displayName = 'Listbox.Option';
var _default = ListboxOption;
exports.default = _default;
